vrpn_Laputa.C

```diff
@@ -0,0 +1,194 @@
+    vector[0] = s.si = (p.i >> 3) & 0x001fffff;
+
+    /* Assemble the vector's y component: */
+    p.b[0] = raw[5];
+    p.b[1] = raw[4];
+    p.b[2] = raw[3];
+    p.b[3] = raw[2];
+    vector[1] = s.si = (p.i >> 6) & 0x001fffff;
+
+    /* Assemble the vector's z component: */
+    p.b[0] = raw[7];
+    p.b[1] = raw[6];
+    p.b[2] = raw[5];
+    // p.b[3]=0U; // Not needed because it's masked out below anyway
+    vector[2] = s.si = (p.i >> 1) & 0x001fffff;
+}
+
+void vrpn_Laputa::parse_message_type_1(std::size_t bytes, vrpn_uint8 *buffer)
+{
+    size_t num_reports = buffer[1];
+    if (num_reports > 3) {
+        num_reports = 3;
+    }
+
+    // Skip past the report type and num_reports bytes and
+    // start parsing there.
+    vrpn_uint8 *bufptr = &buffer[2];
+
+    // The next two bytes are an increasing counter that changes by 1 for
+    // every report.
+    vrpn_uint16 report_index;
+    report_index =
+        vrpn_unbuffer_from_little_endian<vrpn_uint16, vrpn_uint8>(bufptr);
+    channel[1] = report_index;
+
+    // The next two bytes are zero, so we skip them
+    vrpn_uint16 skip;
+    skip = vrpn_unbuffer_from_little_endian<vrpn_uint16, vrpn_uint8>(bufptr);
+
+    // The next entry is temperature, and it may be in hundredths of a degree C
+    vrpn_uint16 temperature;
+    const double temperature_scale = 0.01;
+    temperature =
+        vrpn_unbuffer_from_little_endian<vrpn_uint16, vrpn_uint8>(bufptr);
+    channel[0] = temperature * temperature_scale;
+
+    // The magnetometer data comes after the space to store three
+    // reports.
+    vrpn_uint8 *magnetometer_ptr = &buffer[56];
+    vrpn_int16 magnetometer_raw[3];
+    for (size_t i = 0; i < 3; i++) {
+        magnetometer_raw[i] =
+            vrpn_unbuffer_from_little_endian<vrpn_int16, vrpn_uint8>(
+                magnetometer_ptr);
+    }
+    // Invert all axes to make the magnetometer direction match
+    // the sign of the gravity vector.
+    const double magnetometer_scale = 0.0001;
+    channel[8] = -magnetometer_raw[0] * magnetometer_scale;
+    channel[9] = -magnetometer_raw[1] * magnetometer_scale;
+    channel[10] = -magnetometer_raw[2] * magnetometer_scale;
+
+    // Unpack a 16-byte accelerometer/gyro report using the routines from
+    // Oliver's code.
+    for (size_t i = 0; i < num_reports; i++) {
+        vrpn_int32 accelerometer_raw[3];
+        vrpn_int32 gyroscope_raw[3];
+        unpackVector(bufptr, accelerometer_raw);
+        bufptr += 8;
+        unpackVector(bufptr, gyroscope_raw);
+        bufptr += 8;
+
+        // Compute the double values using default calibration.
+        // The accelerometer data goes into analogs 0,1,2.
+        // The gyro data goes into analogs 3,4,5.
+        // The magnetomoter data goes into analogs 6,7,8.
+        const double accelerometer_scale = 0.0001;
+        const double gyroscope_scale = 0.0001;
+        channel[2] = accelerometer_raw[0] * accelerometer_scale;
```

Comment:
don't know, would probably be safest to `static_cast<double>()` around the `accelerometer_raw` parts